components/brave_news/browser/brave_news_controller.cc

```diff
@@ -455,39 +459,86 @@ void BraveNewsController::GetImageData(const GURL& padded_image_url,
   const std::string ending = ".pad";
   if (file_name.length() >= file_name.max_size() - 1 ||
       file_name.length() <= ending.length()) {
-    std::optional<std::vector<uint8_t>> args;
-    std::move(callback).Run(std::move(args));
+    std::move(callback).Run(std::nullopt);
     return;
   }
   const bool is_padded =
       (file_name.compare(file_name.length() - ending.length(), ending.length(),
                          ending) == 0);
-  VLOG(3) << "is padded: " << is_padded;
+  DVLOG(4) << "is padded: " << is_padded;
+  // Ensure we don't de-pad non PCDN images
+  if (is_padded && !padded_image_url.DomainIs(GetMatchingPCDNHostname())) {
+    DLOG(ERROR) << "Encountered a padded URL apparently not from PCDN: "
+                << padded_image_url.spec();
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
   // Make the request
   private_cdn_request_helper_.DownloadToString(
       padded_image_url,
+      base::BindOnce(&BraveNewsController::OnGetImageDataFetched,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback),
+                     is_padded));
+}
+
+void BraveNewsController::OnGetImageDataFetched(GetImageDataCallback callback,
+                                                const bool is_padded,
+                                                const int response_code,
+                                                const std::string& body) {
+  // Handle the response
+  DVLOG(4) << "getimagedata response code: " << response_code;
+  // Attempt to remove byte padding if applicable
+  std::string_view body_payload(body.data(), body.size());
+  if (response_code != 200 ||
+      (is_padded &&
+       !brave::PrivateCdnHelper::GetInstance()->RemovePadding(&body_payload))) {
+    // Byte padding removal failed
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+  // Download (and optional unpadding) was successful
+  std::vector<uint8_t> image_bytes(body_payload.begin(), body_payload.end());
+#if BUILDFLAG(IS_ANDROID)
```

Comment:
It's not clear why Android is a special case.

Reply:
Only desktop uses WebUI for Brave News. Android uses native UI. Was I correct in assuming the security necessity is only for WebUI?

Reply:
> Was I correct in assuming the security necessity is only for WebUI
AFAIK, there is no rule-of-2 exclusions for Android.
For the native UI the question is what decoder is used. If it's the same C++ decoder (maybe via some Java wrapper), it's not clear why it's safer that for desktop. If it's a decoder purely written on a safe language (i.e. Java), it's okay.

@fmarier WDYT?

Reply:
Removed the condition for android

Reply:
I'm not sure how this is implemented on Android, but the important thing here is that these images coming from third-party servers could be malicious and take over the renderer in which they appear.

On desktop, that would be bad given it's a privileged page.

Reply:
@deeppandya can you let us know how the brave news images are rendered for Android please?

Reply:
we are using https://github.com/brave/brave-core/blob/9b8418eef05eeca82c78a924b47f1f12f20c40db/android/java/org/chromium/chrome/browser/brave_news/CardBuilderFeedCard.java#L1411 to get the image for the card here and we use Glide to rendhttps://github.com/brave/brave-core/blob/9b8418eef05eeca82c78a924b47f1f12f20c40db/android/java/org/chromium/chrome/browser/brave_news/CardBuilderFeedCard.java#L1421er the image here :

Reply:
@petemill Android needs re-encoding too. While Glide is written in Java and uses Java's classes like BitmapFactory, they are just wrappers around C++ classes. So it's not really memory safe.